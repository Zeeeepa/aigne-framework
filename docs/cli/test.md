# Testing

Use AIGNE CLI to run agent tests.

## `aigne test` Command

Run tests in the specified agent directory.

```bash
aigne test [options]
```

### Options

* `--path <path>`: Path to the agent directory (defaults to current directory `.`)
* `--help`: Show command help

### Examples

#### Run tests in current directory

```bash
aigne test
```

#### Run tests in specific directory

```bash
aigne test --path ./my-agents
```

### Test File Structure

The AIGNE testing system automatically finds and runs test files in your project. Make sure your test files follow the correct naming conventions and structure.

### Best Practices

1. **Organize test files**: Place test files in appropriate directory structures
2. **Write clear test cases**: Ensure each test case has a clear purpose and expected outcome
3. **Use descriptive test names**: Make test names clearly express what the test covers
4. **Run tests regularly**: Run tests regularly during development to ensure code quality
