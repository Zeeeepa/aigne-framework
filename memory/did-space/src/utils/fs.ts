export async function streamToString(
  stream: ReadableStream | NodeJS.ReadableStream,
  encoding: BufferEncoding = "utf-8",
): Promise<string> {
  const chunks = [];
  for await (const chunk of stream) {
    chunks.push(chunk);
  }
  return Buffer.concat(chunks).toString(encoding);
}
