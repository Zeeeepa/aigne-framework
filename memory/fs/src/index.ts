export * from "./memory.js";
export * from "./utils/index.js";
