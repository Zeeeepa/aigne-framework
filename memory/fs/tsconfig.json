{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "paths": {
      "@aigne/fs-memory": ["./src"]
    }
  },
  "exclude": ["node_modules", "lib"],
  "include": ["src", "test"]
}
