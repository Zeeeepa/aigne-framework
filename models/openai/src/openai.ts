import OpenAI, { type APIError } from "openai";

// Use a custom OpenAI client to handle API errors for better error messages
export class CustomOpenAI extends OpenAI {
  protected override makeStatusError(
    status: number,
    error: object,
    message: string | undefined,
    headers: Headers,
  ): APIError {
    if (!("error" in error) || typeof error.error !== "string") {
      message = JSON.stringify(error);
    }

    return super.makeStatusError(status, error, message, headers);
  }
}
