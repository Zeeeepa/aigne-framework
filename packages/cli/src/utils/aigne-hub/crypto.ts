import crypto from "node:crypto";
import { AesCrypter } from "@ocap/mcrypto/lib/crypter/aes-legacy.js";

const aes = new AesCrypter();
export const decrypt = (m: string, s: string, i: string) =>
  aes.decrypt(m, crypto.pbkdf2Sync(i, s, 256, 32, "sha512").toString("hex"));
export const encrypt = (m: string, s: string, i: string) =>
  aes.encrypt(m, crypto.pbkdf2Sync(i, s, 256, 32, "sha512").toString("hex"));

const escapeFn = (str: string) => str.replace(/\+/g, "-").replace(/\//g, "_").replace(/=/g, "");
const unescapeFn = (str: string) =>
  (str + "===".slice((str.length + 3) % 4)).replace(/-/g, "+").replace(/_/g, "/");
export const encodeEncryptionKey = (key: string) => escapeFn(Buffer.from(key).toString("base64"));
export const decodeEncryptionKey = (str: string) =>
  new Uint8Array(Buffer.from(unescapeFn(str), "base64"));
