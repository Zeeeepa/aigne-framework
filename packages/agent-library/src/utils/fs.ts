import { stat } from "node:fs/promises";
import { homedir } from "node:os";
import { join } from "node:path";

export async function exists(path: string): Promise<boolean> {
  try {
    await stat(path);
    return true;
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") {
      return false;
    }
    throw e;
  }
}

export function expandHome(filepath: string): string {
  if (filepath.startsWith("~/") || filepath === "~") {
    return join(homedir(), filepath.slice(1));
  }
  return filepath;
}
