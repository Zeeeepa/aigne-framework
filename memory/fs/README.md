# @aigne/fs-memory

<p align="center">
  <picture>
    <source srcset="https://raw.githubusercontent.com/AIGNE-io/aigne-framework/main/logo-dark.svg" media="(prefers-color-scheme: dark)">
    <source srcset="https://raw.githubusercontent.com/AIGNE-io/aigne-framework/main/logo.svg" media="(prefers-color-scheme: light)">
    <img src="https://raw.githubusercontent.com/AIGNE-io/aigne-framework/main/logo.svg" alt="AIGNE Logo" width="400" />
  </picture>
</p>

[![GitHub star chart](https://img.shields.io/github/stars/AIGNE-io/aigne-framework?style=flat-square)](https://star-history.com/#AIGNE-io/aigne-framework)
[![Open Issues](https://img.shields.io/github/issues-raw/AIGNE-io/aigne-framework?style=flat-square)](https://github.com/AIGNE-io/aigne-framework/issues)
[![codecov](https://codecov.io/gh/AIGNE-io/aigne-framework/graph/badge.svg?token=DO07834RQL)](https://codecov.io/gh/AIGNE-io/aigne-framework)
[![NPM Version](https://img.shields.io/npm/v/@aigne/fs-memory)](https://www.npmjs.com/package/@aigne/fs-memory)
[![Elastic-2.0 licensed](https://img.shields.io/npm/l/@aigne/fs-memory)](https://github.com/AIGNE-io/aigne-framework/blob/main/LICENSE)

File system memory component for [AIGNE Framework](https://github.com/AIGNE-io/aigne-framework), providing file system-based memory storage capabilities.

## Introduction

`@aigne/fs-memory` is the file system memory component of [AIGNE Framework](https://github.com/AIGNE-io/aigne-framework), providing local file system-based memory storage and retrieval functionality. This component stores memories as YAML files in specified directories, providing a simple and reliable memory persistence solution.

## Features

* **File System Storage**: Uses local file system for memory file storage
* **YAML Format**: Uses YAML format for memory storage, making it easy to read and edit
* **Directory Management**: Automatically creates and manages memory directory structure
* **Path Support**: Supports absolute paths, relative paths, and home directory (~) expansion
* **AI Agent Management**: Uses AI agents to handle memory recording and retrieval
* **TypeScript Support**: Complete type definitions providing an excellent development experience

## Installation

### Using npm

```bash
npm install @aigne/fs-memory
```

### Using yarn

```bash
yarn add @aigne/fs-memory
```

### Using pnpm

```bash
pnpm add @aigne/fs-memory
```

## Basic Usage

```typescript
import { AIAgent, AIGNE } from "@aigne/core";
import { FSMemory } from "@aigne/fs-memory";
import { OpenAIChatModel } from "@aigne/openai";

// Create AI model instance
const model = new OpenAIChatModel({
  apiKey: process.env.OPENAI_API_KEY,
  model: "gpt-4-turbo",
});

// Create file system memory
const memory = new FSMemory({
  rootDir: "./memories",
});

// Create AI agent with file-based memory
const agent = AIAgent.from({
  name: "FileAssistant",
  instructions: "You are a helpful assistant with file-based memory.",
  memory: memory,
  inputKey: "message",
});

// Use AIGNE execution engine
const aigne = new AIGNE({ model });

// Invoke agent
const userAgent = await aigne.invoke(agent);

// Send message
const response = await userAgent.invoke({
  message: "Remember that I prefer working in the morning",
});

console.log(response.message);

// Query memory
const response2 = await userAgent.invoke({
  message: "What do you know about my work preferences?",
});

console.log(response2.message);
```

## Advanced Configuration

### Custom Memory Directory

```typescript
import { join } from "node:path";
import { FSMemory } from "@aigne/fs-memory";

// Using absolute path
const memory1 = new FSMemory({
  rootDir: "/absolute/path/to/memories",
});

// Using relative path
const memory2 = new FSMemory({
  rootDir: "./data/memories",
});

// Using home directory
const memory3 = new FSMemory({
  rootDir: "~/my-app/memories",
});

// Using dynamic path
const memory4 = new FSMemory({
  rootDir: join(process.cwd(), "memories"),
});
```

### Configure Memory Agent Options

```typescript
const memory = new FSMemory({
  rootDir: "./memories",
  // Custom memory retriever agent configuration
  retrieverOptions: {
    name: "CustomRetriever",
    instructions: "Custom retrieval instructions for file-based memory",
  },
  // Custom memory recorder agent configuration
  recorderOptions: {
    name: "CustomRecorder",
    instructions: "Custom recording instructions for file-based memory",
  },
  // Other agent configurations
  autoUpdate: true,
});
```

## Using Multiple Memory Types Together

```typescript
import { DefaultMemory } from "@aigne/default-memory";
import { FSMemory } from "@aigne/fs-memory";

// Combine multiple memory types
const agent = AIAgent.from({
  name: "MultiMemoryAgent",
  instructions: "You are an assistant with multiple memory types.",
  memory: [
    // Database memory for structured data
    new DefaultMemory({
      storage: {
        url: "file:memory.db",
      },
    }),
    // File system memory for long-term storage
    new FSMemory({
      rootDir: "./long-term-memories",
    }),
  ],
  inputKey: "message",
});
```

## Memory File Structure

FSMemory creates the following structure in the specified directory:

```
memories/
├── conversation-id-1/
│   └── memory.yaml
├── conversation-id-2/
│   └── memory.yaml
└── ...
```

Each memory file contains YAML-formatted memory data:

```yaml
- id: memory-id-1
  content: "User prefers working in the morning"
  timestamp: 2024-01-01T10:00:00Z
  metadata:
    type: "preference"

- id: memory-id-2
  content: "User is a software developer"
  timestamp: 2024-01-01T10:05:00Z
  metadata:
    type: "personal_info"
```

## License

Elastic-2.0
