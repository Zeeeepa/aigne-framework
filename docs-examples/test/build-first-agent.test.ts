import { expect, mock, spyOn, test } from "bun:test";
import assert from "node:assert";
import { randomUUID } from "node:crypto";
import { mkdir, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { AIAgent, AIGNE, FunctionAgent, MCPAgent } from "@aigne/core";
import { OpenAIChatModel } from "@aigne/openai";
import { AIGNEHTTPClient } from "@aigne/transport/http-client/index.js";
import { AIGNEHTTPServer } from "@aigne/transport/http-server/index.js";
import detectPort from "detect-port";
import express from "express";

test("Build first agent: basic", async () => {
  // #region example-build-first-agent

  // #region example-build-first-agent-create-aigne
  const aigne = new AIGNE({
    model: new OpenAIChatModel({
      apiKey: process.env.OPENAI_API_KEY,
      model: "gpt-4o-mini",
    }),
  });
  // #endregion example-build-first-agent-create-aigne
  assert(aigne.model);

  // #region example-build-first-agent-create-agent
  const agent = AIAgent.from({
    instructions: "You are a helpful assistant for Crypto market cap",
  });
  // #endregion example-build-first-agent-create-agent

  spyOn(aigne.model, "process").mockReturnValueOnce({
    text: "Cryptocurrency, often referred to as crypto, is a type of digital or virtual currency that uses cryptography for security",
  });

  // #region example-build-first-agent-invoke-agent
  const result = await aigne.invoke(agent, "What is crypto?");
  console.log(result);
  // Output: { $message: "Cryptocurrency, often referred to as crypto, is a type of digital or virtual currency that uses cryptography for security" }
  // #endregion example-build-first-agent-invoke-agent

  expect(result).toEqual({
    $message:
      "Cryptocurrency, often referred to as crypto, is a type of digital or virtual currency that uses cryptography for security",
  });

  // #endregion example-build-first-agent
});

test("Build first agent: MCP server as Agent", async () => {
  // #region example-mcp-server-as-agent

  // #region example-mcp-server-as-agent-create-agent
  spyOn(MCPAgent, "from").mockReturnValueOnce(
    new MCPAgent({
      name: "ccxt",
      client: mock() as unknown as MCPAgent["client"],
      skills: [
        FunctionAgent.from({
          name: "get-ticker",
          process: () => ({}),
        }),
      ],
    }),
  );

  const ccxt = await MCPAgent.from({
    command: "npx",
    args: ["-y", "@mcpfun/mcp-server-ccxt"],
  });
  // #endregion example-mcp-server-as-agent-create-agent

  // #region example-mcp-server-as-agent-explore-skills
  console.log(ccxt.skills);
  // Output: [ cache-stats, clear-cache, set-log-level, list-exchanges, get-ticker, batch-get-tickers, get-orderbook, get-ohlcv, get-trades, get-markets, get-exchange-info, get-leverage-tiers, get-funding-rates, get-market-types, account-balance, place-market-order, set-leverage, set-margin-mode, place-futures-market-order, get-proxy-config, set-proxy-config, test-proxy-connection, clear-exchange-cache, set-market-type ]

  const getTicker = ccxt.skills["get-ticker"];
  assert(getTicker);
  // #endregion example-mcp-server-as-agent-explore-skills

  spyOn(getTicker, "process").mockReturnValueOnce({
    content: [
      {
        type: "text",
        text: '{\n  "symbol": "ABT/USD",\n  "timestamp": 1747789089514,\n  "datetime": "2025-05-21T00:58:09.514083Z",\n  "bid": 0.9336,\n  "ask": 0.935,\n  "last": 0.9338,\n  "close": 0.9338,\n  "info": {\n    "trade_id": "5572965",\n    "product_id": "ABT-USD",\n    "price": "0.9338",\n    "size": "17",\n    "time": "2025-05-21T00:58:09.514083Z",\n    "side": "BUY",\n    "bid": "",\n    "ask": "",\n    "exchange": "coinbase"\n  }\n}',
      },
    ],
  });

  // #region example-mcp-server-as-agent-invoke-skill
  const result = await getTicker.invoke({ exchange: "coinbase", symbol: "ABT/USD" });
  console.log(result);
  // Output: { content: [ { type: 'text', text: '{\n  "symbol": "ABT/USD",\n  "timestamp": 1747789089514,\n  "datetime": "2025-05-21T00:58:09.514083Z",\n  "bid": 0.9336,\n  "ask": 0.935,\n  "last": 0.9338,\n  "close": 0.9338,\n  "info": {\n    "trade_id": "5572965",\n    "product_id": "ABT-USD",\n    "price": "0.9338",\n    "size": "17",\n    "time": "2025-05-21T00:58:09.514083Z",\n    "side": "BUY",\n    "bid": "",\n    "ask": "",\n    "exchange": "coinbase"\n  }\n}' } ] }
  expect(result).toEqual({
    content: [
      {
        type: "text",
        text: '{\n  "symbol": "ABT/USD",\n  "timestamp": 1747789089514,\n  "datetime": "2025-05-21T00:58:09.514083Z",\n  "bid": 0.9336,\n  "ask": 0.935,\n  "last": 0.9338,\n  "close": 0.9338,\n  "info": {\n    "trade_id": "5572965",\n    "product_id": "ABT-USD",\n    "price": "0.9338",\n    "size": "17",\n    "time": "2025-05-21T00:58:09.514083Z",\n    "side": "BUY",\n    "bid": "",\n    "ask": "",\n    "exchange": "coinbase"\n  }\n}',
      },
    ],
  });
  // #endregion example-mcp-server-as-agent-invoke-skill

  // #endregion example-mcp-server-as-agent
});

test("Build first agent: add skills to agent", async () => {
  spyOn(MCPAgent, "from").mockReturnValueOnce(
    new MCPAgent({
      name: "ccxt",
      client: mock() as unknown as MCPAgent["client"],
      skills: [
        FunctionAgent.from({
          name: "get-ticker",
          process: () => ({}),
        }),
      ],
    }),
  );

  // #region example-add-skills-to-agent

  const aigne = new AIGNE({
    model: new OpenAIChatModel(),
  });
  assert(aigne.model);

  // #region example-add-skills-to-agent-create-skill-agent
  const ccxt = await MCPAgent.from({
    command: "npx",
    args: ["-y", "@mcpfun/mcp-server-ccxt"],
  });
  // #endregion example-add-skills-to-agent-create-skill-agent

  // #region example-add-skills-to-agent-add-skills
  const agent = AIAgent.from({
    instructions: "You are a helpful assistant for Crypto market cap",
    skills: [ccxt],
  });
  // #endregion example-add-skills-to-agent-add-skills

  spyOn(aigne.model, "process").mockReturnValueOnce({
    text: "The current price of ABT/USD on Coinbase is $0.9684.",
  });

  // #region example-add-skills-to-agent-invoke-agent
  const result = await aigne.invoke(agent, "What is the crypto price of ABT/USD on coinbase?");
  console.log(result);
  // Output: { $message:"The current price of ABT/USD on Coinbase is $0.9684." }
  expect(result).toEqual({ $message: "The current price of ABT/USD on Coinbase is $0.9684." });
  // #endregion example-add-skills-to-agent-invoke-agent

  // #endregion example-add-skills-to-agent
});

test("Build first agent: enable memory for agent", async () => {
  const tmp = join(tmpdir(), randomUUID());
  const memoryStoragePath = join(tmp, "memory.db");
  await mkdir(tmp, { recursive: true });

  // #region example-enable-memory-for-agent

  const aigne = new AIGNE({
    model: new OpenAIChatModel(),
  });
  assert(aigne.model);

  // #region example-enable-memory-for-agent-enable-memory
  const agent = AIAgent.from({
    instructions: "You are a helpful assistant for Crypto market analysis",
    memory: {
      storage: {
        path: memoryStoragePath, // Path to store memory data, such as './memory.db'
      },
    },
  });
  // #endregion example-enable-memory-for-agent-enable-memory

  spyOn(aigne.model, "process")
    .mockReturnValueOnce({
      text: "Nice to meet you, John Doe! Bitcoin is an interesting cryptocurrency to invest in. How long have you been investing in crypto? Do you have a diversified portfolio?",
    })
    .mockReturnValueOnce({
      text: "Your favorite cryptocurrency is Bitcoin.",
    })
    .mockReturnValueOnce({
      text: "Got it, you've invested $5000 in Ethereum! That's a good investment. If there's anything else you'd like to share about your crypto portfolio or have questions, feel free!",
    })
    .mockReturnValueOnce({
      text: "You've invested $5000 in Ethereum.",
    });

  // #region example-enable-memory-for-agent-invoke-agent-1
  const result1 = await aigne.invoke(agent, "My name is John Doe and I like to invest in Bitcoin.");
  console.log(result1);
  // Output: { $message: "Nice to meet you, John Doe! Bitcoin is an interesting cryptocurrency to invest in. How long have you been investing in crypto? Do you have a diversified portfolio?" }
  expect(result1).toEqual({
    $message:
      "Nice to meet you, John Doe! Bitcoin is an interesting cryptocurrency to invest in. How long have you been investing in crypto? Do you have a diversified portfolio?",
  });
  // #endregion example-enable-memory-for-agent-invoke-agent-1

  // #region example-enable-memory-for-agent-invoke-agent-2
  const result2 = await aigne.invoke(agent, "What is my favorite cryptocurrency?");
  console.log(result2);
  // Output: { $message: "Your favorite cryptocurrency is Bitcoin." }
  expect(result2).toEqual({ $message: "Your favorite cryptocurrency is Bitcoin." });
  // #endregion example-enable-memory-for-agent-invoke-agent-2

  // #region example-enable-memory-for-agent-invoke-agent-3
  const result3 = await aigne.invoke(agent, "I've invested $5000 in Ethereum.");
  console.log(result3);
  // Output: { $message: "Got it, you've invested $5000 in Ethereum! That's a good investment. If there's anything else you'd like to share about your crypto portfolio or have questions, feel free!" }
  expect(result3).toEqual({
    $message:
      "Got it, you've invested $5000 in Ethereum! That's a good investment. If there's anything else you'd like to share about your crypto portfolio or have questions, feel free!",
  });
  // #endregion example-enable-memory-for-agent-invoke-agent-3

  // #region example-enable-memory-for-agent-invoke-agent-4
  const result4 = await aigne.invoke(agent, "How much have I invested in Ethereum?");
  console.log(result4);
  // Output: { $message: "You've invested $5000 in Ethereum." }
  expect(result4).toEqual({ $message: "You've invested $5000 in Ethereum." });
  // #endregion example-enable-memory-for-agent-invoke-agent-4

  // #endregion example-enable-memory-for-agent

  await rm(tmp, { recursive: true, force: true });
});

test("Build first agent: custom user context", async () => {
  const tmp = join(tmpdir(), randomUUID());
  const memoryStoragePath = join(tmp, "memory.db");
  await mkdir(tmp, { recursive: true });

  // #region example-custom-user-context

  const aigne = new AIGNE({
    model: new OpenAIChatModel(),
  });
  assert(aigne.model);

  // #region example-custom-user-context-create-agent
  const agent = AIAgent.from({
    instructions: "You are a helpful assistant for Crypto market analysis",
    memory: {
      storage: {
        path: memoryStoragePath, // Path to store memory data, such as './memory.db'
        getSessionId: ({ userContext }) => userContext.userId as string, // Use userId from userContext as session ID
      },
    },
  });
  // #endregion example-custom-user-context-create-agent

  // #region example-custom-user-context-invoke-agent
  spyOn(aigne.model, "process").mockReturnValueOnce({
    text: "Nice to meet you, John Doe! Bitcoin is an interesting cryptocurrency to invest in. How long have you been investing in crypto? Do you have a diversified portfolio?",
  });
  const result = await aigne.invoke(agent, "My name is John Doe and I like to invest in Bitcoin.", {
    userContext: { userId: "user_123" },
  });
  console.log(result);
  // Output: { $message: "Nice to meet you, John Doe! Bitcoin is an interesting cryptocurrency to invest in. How long have you been investing in crypto? Do you have a diversified portfolio?" }
  expect(result).toEqual({
    $message:
      "Nice to meet you, John Doe! Bitcoin is an interesting cryptocurrency to invest in. How long have you been investing in crypto? Do you have a diversified portfolio?",
  });
  // #endregion example-custom-user-context-invoke-agent

  // #endregion example-custom-user-context

  await rm(tmp, { recursive: true, force: true });
});

test("Build first agent: serve agent as API service", async () => {
  const tmp = join(tmpdir(), randomUUID());
  const memoryStoragePath = join(tmp, "memory.db");
  await mkdir(tmp, { recursive: true });

  // #region example-serve-agent-as-api-service

  // #region example-serve-agent-as-api-service-create-named-agent
  const agent = AIAgent.from({
    name: "chatbot",
    instructions: "You are a helpful assistant",
    memory: {
      storage: {
        path: memoryStoragePath, // Path to store memory data, such as './memory.db'
        getSessionId: ({ userContext }) => userContext.userId as string, // Use userId from userContext as session ID
      },
    },
  });
  // #endregion example-serve-agent-as-api-service-create-named-agent

  // #region example-serve-agent-as-api-service-create-aigne
  const aigne = new AIGNE({
    model: new OpenAIChatModel(),
    agents: [agent],
  });
  assert(aigne.model);
  // #endregion example-serve-agent-as-api-service-create-aigne

  // #region example-serve-agent-as-api-service-create-http-server
  const server = new AIGNEHTTPServer(aigne);

  const app = express();

  app.post("/api/chat", async (req, res) => {
    const userId = "user_123"; // Example user ID, replace with actual logic to get user ID, such as `req.user.id` in a real application
    await server.invoke(req, res, { userContext: { userId } });
  });

  let port = 3000;
  port = await detectPort();

  const httpServer = app.listen(port);
  // #endregion example-serve-agent-as-api-service-create-http-server

  // #endregion example-serve-agent-as-api-service

  // #region example-aigne-http-client-usage

  // #region example-aigne-http-client-create-client
  const client = new AIGNEHTTPClient({ url: `http://localhost:${port}/api/chat` });
  // #endregion example-aigne-http-client-create-client

  spyOn(aigne.model, "process").mockReturnValueOnce({
    text: "The current price of ABT/USD on Coinbase is $0.9684.",
  });

  // #region example-aigne-http-client-invoke-agent
  const result = await client.invoke("chatbot", "What is the crypto price of ABT/USD on coinbase?");
  console.log(result);
  // Output: { $message: "The current price of ABT/USD on Coinbase is $0.9684." }
  expect(result).toEqual({ $message: "The current price of ABT/USD on Coinbase is $0.9684." });
  // #endregion example-aigne-http-client-invoke-agent

  // #endregion example-aigne-http-client-usage

  httpServer.closeAllConnections();
  httpServer.close();

  await rm(tmp, { recursive: true, force: true });
});
