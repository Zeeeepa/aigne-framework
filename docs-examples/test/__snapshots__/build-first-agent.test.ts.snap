// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Build first agent: setup memory for client agent 1`] = `
[
  {
    "messages": [
      {
        "content": "You are a helpful assistant",
        "name": undefined,
        "role": "system",
      },
      {
        "content": "What is the crypto price of ABT/USD on coinbase?",
        "role": "user",
      },
      {
        "content": "The current price of ABT/USD on Coinbase is $0.9684.",
        "role": "agent",
      },
      {
        "content": "What question did I just ask?",
        "role": "user",
      },
    ],
    "modelOptions": undefined,
    "responseFormat": undefined,
    "toolChoice": undefined,
    "tools": undefined,
  },
  Anything,
]
`;
